linkerd/proxy/http/src/orig_proto.rs

```diff
@@ -26,93 +27,85 @@ pub struct Downgrade<S> {
+        }
 
-        let version = req.version();
+        let orig_version = req.version();
+        let absolute_form = req
+            .extensions_mut()
+            .remove::<h1::WasAbsoluteForm>()
+            .is_some();
+        debug!(version = ?orig_version, absolute_form, "Upgrading request");
 
         // absolute-form is far less common, origin-form is the usual,
         // so only encode the extra information if it's different than
         // the normal.
-        let val = match (version, self.absolute_form) {
+        let header = match (orig_version, absolute_form) {
             (http::Version::HTTP_11, false) => "HTTP/1.1",
             (http::Version::HTTP_11, true) => "HTTP/1.1; absolute-form",
             (http::Version::HTTP_10, false) => "HTTP/1.0",
             (http::Version::HTTP_10, true) => "HTTP/1.0; absolute-form",
             (v, _) => unreachable!("bad orig-proto version: {:?}", v),
         };
-        debug!("Upgrading request to HTTP2 from {}", val);
         req.headers_mut()
-            .insert(L5D_ORIG_PROTO, HeaderValue::from_static(val));
+            .insert(L5D_ORIG_PROTO, HeaderValue::from_static(header));
 
         // transfer-encoding is illegal in HTTP2
         req.headers_mut().remove(TRANSFER_ENCODING);
 
         *req.version_mut() = http::Version::HTTP_2;
 
-        UpgradeFuture {
-            version,
-            inner: self.inner.call(req),
-        }
-    }
-}
-
-#[pin_project]
-pub struct UpgradeFuture<F> {
-    version: http::Version,
-    #[pin]
-    inner: F,
-}
-
-impl<F, B> Future for UpgradeFuture<F>
-where
-    F: TryFuture<Ok = http::Response<B>>,
-{
-    type Output = Result<F::Ok, F::Error>;
-
-    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
-        let this = self.project();
-        let mut res = ready!(this.inner.try_poll(cx))?;
-        let version = res
-            .headers_mut()
-            .remove(L5D_ORIG_PROTO)
-            .and_then(|orig_proto| {
-                if orig_proto == "HTTP/1.1" {
-                    Some(http::Version::HTTP_11)
-                } else if orig_proto == "HTTP/1.0" {
-                    Some(http::Version::HTTP_10)
-                } else {
-                    None
-                }
-            })
-            .unwrap_or(*this.version);
-        debug!("Downgrading response to {:?}", version);
-        *res.version_mut() = version;
-        Poll::Ready(Ok(res.into()))
+        Box::pin(self.h2.call(req).map_ok(move |mut rsp| {
+            let version = rsp
+                .headers_mut()
+                .remove(L5D_ORIG_PROTO)
+                .and_then(|orig_proto| {
+                    if orig_proto == "HTTP/1.1" {
+                        Some(http::Version::HTTP_11)
+                    } else if orig_proto == "HTTP/1.0" {
+                        Some(http::Version::HTTP_10)
+                    } else {
+                        None
+                    }
```

Comment:
nit/tioli: could be 
```suggestion
                    match orig_proto {
                        "HTTP/1.1" =>  Some(http::Version::HTTP_11),
                        "HTTP/1.0" => Some(http::Version::HTTP_10),
                        _ => None,
                    }
```
...unless i'm missing something, of course?

Reply:
`orig_proto` is a HeaderValue and not a string, so the matching form doesn't work